File: backend/src/main/java/com/chatapp/service/ChatService.java
package com.chatapp.service;

import com.chatapp.model.Chat;
import com.chatapp.model.User;
import com.chatapp.repository.ChatRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ChatService {

    private final ChatRepository chatRepository;
    private final UserService userService;

    public Chat createPrivateChat(String otherUserId) {
        User currentUser = userService.getCurrentUser();
        User otherUser = userService.getUserById(otherUserId);

        // Verificar si ya existe un chat privado entre estos usuarios
        Optional<Chat> existingChat = chatRepository.findPrivateChatBetweenUsers(
                currentUser.getId(), otherUserId);

        if (existingChat.isPresent()) {
            return existingChat.get();
        }

        Chat chat = Chat.builder()
                .type(Chat.ChatType.PRIVATE)
                .participantIds(List.of(currentUser.getId(), otherUserId))
                .createdBy(currentUser.getId())
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .settings(Chat.ChatSettings.builder()
                        .notificationsEnabled(true)
                        .customNotificationTone("default")
                        .isArchived(false)
                        .isPinned(false)
                        .build())
                .build();

        return chatRepository.save(chat);
    }

    public Chat createGroupChat(String name, String description, List<String> participantIds) {
        User currentUser = userService.getCurrentUser();

        // Agregar al creador como participante
        if (!participantIds.contains(currentUser.getId())) {
            participantIds.add(currentUser.getId());
        }

        Chat chat = Chat.builder()
                .type(Chat.ChatType.GROUP)
                .name(name)
                .description(description)
                .participantIds(participantIds)
                .createdBy(currentUser.getId())
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .settings(Chat.ChatSettings.builder()
                        .notificationsEnabled(true)
                        .customNotificationTone("default")
                        .isArchived(false)
                        .isPinned(false)
                        .build())
                .build();

        return chatRepository.save(chat);
    }

    public List<Chat> getUserChats() {
        User currentUser = userService.getCurrentUser();
        return chatRepository.findByParticipantId(currentUser.getId());
    }

    public Optional<Chat> getChatById(String chatId) {
        return chatRepository.findById(chatId);
    }

    public Chat updateChatSettings(String chatId, Chat.ChatSettings settings) {
        Chat chat = chatRepository.findById(chatId)
                .orElseThrow(() -> new RuntimeException("Chat no encontrado"));

        chat.setSettings(settings);
        chat.setUpdatedAt(LocalDateTime.now());

        return chatRepository.save(chat);
    }

    public void archiveChat(String chatId) {
        Chat chat = chatRepository.findById(chatId)
                .orElseThrow(() -> new RuntimeException("Chat no encontrado"));

        chat.getSettings().setIsArchived(true);
        chat.setUpdatedAt(LocalDateTime.now());
        chatRepository.save(chat);
    }

    public void pinChat(String chatId) {
        Chat chat = chatRepository.findById(chatId)
                .orElseThrow(() -> new RuntimeException("Chat no encontrado"));

        chat.getSettings().setIsPinned(true);
        chat.setUpdatedAt(LocalDateTime.now());
        chatRepository.save(chat);
    }

    public void unpinChat(String chatId) {
        Chat chat = chatRepository.findById(chatId)
                .orElseThrow(() -> new RuntimeException("Chat no encontrado"));

        chat.getSettings().setIsPinned(false);
        chat.setUpdatedAt(LocalDateTime.now());
        chatRepository.save(chat);
    }

    public Optional<Chat> getPrivateChatBetweenUsers(String user1Id, String user2Id) {
        return chatRepository.findPrivateChatBetweenUsers(user1Id, user2Id);
    }
}

File: frontend/src/app/core/services/chat.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Chat, ChatSettings } from '../models/chat.model';

@Injectable({
    providedIn: 'root'
})
export class ChatService {
    private apiUrl = `${environment.apiUrl}/chats`;

    constructor(private http: HttpClient) { }

    createPrivateChat(userId: string): Observable<Chat> {
        return this.http.post<Chat>(`${this.apiUrl}/private/${userId}`, {});
    }

    createGroupChat(name: string, description: string, participantIds: string[]): Observable<Chat> {
        return this.http.post<Chat>(`${this.apiUrl}/group`, {
            name,
            description,
            participantIds
        });
    }

    getUserChats(): Observable<Chat[]> {
        return this.http.get<Chat[]>(this.apiUrl);
    }

    getChatById(chatId: string): Observable<Chat> {
        return this.http.get<Chat>(`${this.apiUrl}/${chatId}`);
    }

    updateChatSettings(chatId: string, settings: ChatSettings): Observable<Chat> {
        return this.http.put<Chat>(`${this.apiUrl}/${chatId}/settings`, settings);
    }

    archiveChat(chatId: string): Observable<void> {
        return this.http.post<void>(`${this.apiUrl}/${chatId}/archive`, {});
    }

    pinChat(chatId: string): Observable<void> {
        return this.http.post<void>(`${this.apiUrl}/${chatId}/pin`, {});
    }

    unpinChat(chatId: string): Observable<void> {
        return this.http.post<void>(`${this.apiUrl}/${chatId}/unpin`, {});
    }

    getPrivateChatBetweenUsers(user1Id: string, user2Id: string): Observable<Chat> {
        return this.http.get<Chat>(`${this.apiUrl}/private/${user1Id}/${user2Id}`);
    }
}